File: app/src/main/java/com/xpwnit/dualbt/logging/AppLogger.java
package com.xpwnit.dualbt.logging;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

public final class AppLogger {
    public interface Listener {
        void onLogsChanged();
    }

    private static final String APP_TAG = "DualBT";
    private static final LogStore STORE = new LogStore(500);
    private static final ArrayList<Listener> LISTENERS = new ArrayList<>();
    private static final Handler MAIN = new Handler(Looper.getMainLooper());
    private static FileLogSink fileSink;

    private AppLogger() {
    }

    public static synchronized void initialize(Context context) {
        File logFile = new File(new File(context.getFilesDir(), "logs"), "dualbt.log");
        fileSink = new FileLogSink(logFile, 256 * 1024);
        i("AppLogger", "File logging initialized at " + logFile.getAbsolutePath());
    }

    public static void d(String tag, String message) {
        log("DEBUG", Log.DEBUG, tag, message, null);
    }

    public static void i(String tag, String message) {
        log("INFO", Log.INFO, tag, message, null);
    }

    public static void w(String tag, String message) {
        log("WARN", Log.WARN, tag, message, null);
    }

    public static void e(String tag, String message, Throwable throwable) {
        log("ERROR", Log.ERROR, tag, message, throwable);
    }

    public static List<LogStore.Entry> logs(String level) {
        return STORE.filter(level);
    }

    public static void clear() {
        STORE.clear();
        FileLogSink sink = fileSink;
        if (sink != null) {
            try {
                sink.clear();
            } catch (RuntimeException exception) {
                Log.w(APP_TAG + ":AppLogger", "Unable to clear file log", exception);
            }
        }
        notifyListeners();
    }

    public static void addListener(Listener listener) {
        synchronized (LISTENERS) {
            if (!LISTENERS.contains(listener)) {
                LISTENERS.add(listener);
            }
        }
    }

    public static void removeListener(Listener listener) {
        synchronized (LISTENERS) {
            LISTENERS.remove(listener);
        }
    }

    private static void log(String level, int priority, String tag, String message, Throwable throwable) {
        String fullTag = APP_TAG + ":" + tag;
        if (throwable == null) {
            Log.println(priority, fullTag, message);
        } else {
            Log.e(fullTag, message, throwable);
        }
        LogStore.Entry entry = STORE.add(level, tag, throwable == null ? message : message + " | " + throwable.getMessage());
        writeToFile(entry);
        notifyListeners();
    }

    private static void writeToFile(LogStore.Entry entry) {
        FileLogSink sink = fileSink;
        if (sink == null) {
            return;
        }
        try {
            sink.append(entry);
        } catch (RuntimeException exception) {
            Log.w(APP_TAG + ":AppLogger", "Unable to persist log entry", exception);
        }
    }

    private static void notifyListeners() {
        MAIN.post(() -> {
            List<Listener> copy;
            synchronized (LISTENERS) {
                copy = new ArrayList<>(LISTENERS);
            }
            for (Listener listener : copy) {
                listener.onLogsChanged();
            }
        });
    }
}

File: app/src/main/java/com/xpwnit/dualbt/logging/AppLogger.kt
package com.xpwnit.dualbt.logging

import android.util.Log
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.concurrent.ConcurrentLinkedDeque

enum class LogLevel(val label: String, val priority: Int) {
    DEBUG("DEBUG", Log.DEBUG),
    INFO("INFO", Log.INFO),
    WARNING("WARN", Log.WARN),
    ERROR("ERROR", Log.ERROR)
}

data class LogEntry(
    val timestamp: Long = System.currentTimeMillis(),
    val level: LogLevel,
    val tag: String,
    val message: String,
    val throwable: Throwable? = null
) {
    private val dateFormat = SimpleDateFormat("HH:mm:ss.SSS", Locale.US)

    fun formatted(): String {
        val time = dateFormat.format(Date(timestamp))
        val err = throwable?.let { " | ${it.message}" } ?: ""
        return "[$time] [${level.label}] [$tag] $message$err"
    }
}

object AppLogger {
    private const val MAX_LOG_ENTRIES = 500
    private const val APP_TAG = "DualBT"

    private val _logs = ConcurrentLinkedDeque<LogEntry>()
    private val _logsFlow = MutableStateFlow<List<LogEntry>>(emptyList())
    val logs: StateFlow<List<LogEntry>> = _logsFlow

    var minLevel: LogLevel = LogLevel.DEBUG

    fun d(tag: String, message: String) = log(LogLevel.DEBUG, tag, message)
    fun i(tag: String, message: String) = log(LogLevel.INFO, tag, message)
    fun w(tag: String, message: String) = log(LogLevel.WARNING, tag, message)
    fun e(tag: String, message: String, throwable: Throwable? = null) =
        log(LogLevel.ERROR, tag, message, throwable)

    private fun log(level: LogLevel, tag: String, message: String, throwable: Throwable? = null) {
        if (level.priority < minLevel.priority) return

        // Android Logcat
        val fullTag = "$APP_TAG:$tag"
        when (level) {
            LogLevel.DEBUG -> Log.d(fullTag, message)
            LogLevel.INFO -> Log.i(fullTag, message)
            LogLevel.WARNING -> Log.w(fullTag, message)
            LogLevel.ERROR -> {
                if (throwable != null) Log.e(fullTag, message, throwable)
                else Log.e(fullTag, message)
            }
        }

        // In-memory log store
        val entry = LogEntry(level = level, tag = tag, message = message, throwable = throwable)
        _logs.addFirst(entry)
        while (_logs.size > MAX_LOG_ENTRIES) {
            _logs.removeLast()
        }
        _logsFlow.value = _logs.toList()
    }

    fun clear() {
        _logs.clear()
        _logsFlow.value = emptyList()
    }

    fun getFilteredLogs(level: LogLevel? = null, tag: String? = null): List<LogEntry> {
        return _logs.filter { entry ->
            (level == null || entry.level == level) &&
            (tag == null || entry.tag.contains(tag, ignoreCase = true))
        }
    }
}
